src/js/common/validator.js

```diff
@@ -1,122 +1,470 @@
-        if (that.ajv.getSchema(schemaKey)) {
-            that.ajv.removeSchema(schemaKey);
+    /**
+     *
+     * Strip empty elements from an array.  Used to handle leading delimiters, such as `.path.to.something`, where we
+     * aren't interested in the implicit leading empty string before the first dot.  Only strings and integers are allowed.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {String} item - An item to be evaluated.
+     * @return {Boolean} - True if the item is of non-zero length, false otherwise.
+     */
+    gpii.schema.removeEmptyItems = function (item) { return (typeof item === "string" && item.length) || Number.isInteger(item); };
+
+    /**
+     *
+     * Extract a standardised set of EL path segments that point to a piece of data that breaks a validation rule.  The
+     * handling of "required" fields is standardised to the JSON Schema v3 convention, i.e. the path returned is to
+     * the missing element rather than the parent element that should contain the missing element.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {Object} rawError - A single validation error as returned by AJV.
+     * @return {Array.<String>} - An array of EL path segments representing the path to the invalid or missing material.
+     *
+     */
+    gpii.schema.validator.extractElDataPathSegmentsFromError = function (rawError) {
+        // We have to use this approach so that we can correctly break up dataPath values that contain escaped dots and apostrophes, i.e. `normal.['dotted.field'].alsoNormal`.
+        var rawPathSegments = rawError.dataPath.match(/(([^\[\.]+)|\['([^\]]+)'\])/g) || [];
+        var ajvPathSegments = fluid.transform(rawPathSegments, function (pathSegment) {
+            // Extract the inner value from enclosing brackets and quotes and remove backslashes.
+            return pathSegment.replace(/\['([^\]]+)'\]/, "$1").replace(/\\/g, "");
+        });
+
+        if (rawError.keyword === "required") {
+            ajvPathSegments.push(gpii.schema.validator.trimLeadingDot(rawError.params.missingProperty));
         }
 
-        try {
-            that.ajv.addSchema(schemaContent, schemaKey);
+        return ajvPathSegments;
+    };
+
+
+    /**
+     *
+     * Standardise the path to the failing rule within a GSS schema based on a raw AJV validation error.
+     *
+     * NOTE: This function is a non-API function, i.e. one that assists public functions in performing their work, but
+     * which is not guaranteed to remain available.
+     *
+     * @param {Object} ajvError - A single raw validation error as returned by AJV.
```

Comment:
Can we have more specific eslint type here to explain the contents of the error structure - using a @typedef - (note to ourselves - someday we will have to fold the capabilities of eslint in describing structures together with those of JSON schema itself)

Reply:
This should be a good warmup for sketching out a typedef for the GSS metaschema itself, so I'll give it a go.

Reply:
OK, I've committed a typedef for this.